torch/csrc/distributed/c10d/NCCLUtils.hpp

```diff
@@ -472,6 +472,17 @@ class NCCLComm {
 #endif
   }
 
+  uint64_t getCommHash() {
+#ifdef NCCL_COMM_GET_UNIQUE_HASH
+    // query from NCCL when first time called
+    if (commHash_ == 0) {
+      C10D_NCCL_CHECK(
+          ncclCommGetUniqueHash(ncclComm_, &commHash_), c10::nullopt);
```

Comment:
Unfortunately it might be difficult to accept APIs that's not public. 
We'd like functionalities in PyTorch's OSS code be accessible for all PyTorch users.

There are two solutions:
(1) Advocate and upstream the API to public NCCL;
(2) Establish an abstract interface between PyTorch and a "plugin". In the plugin, you can make this non-public API call.

Reply:
Btw, the value of a ncclComm_t is actually an address pointer. So it should be unique enough and safe to be used as a "hash".

Reply:
> Btw, the value of a ncclComm_t is actually an address pointer (which points to the comm object). So it should be unique enough and safe to be used as a "hash".

Hi @kwen2501 , the pointer is a local ptr, which gives different value across rank. In contrast, we want to expose commHash which is a global unique id used by all ranks in a given NCCL communicator

Reply:
> Unfortunately it might be difficult to accept APIs that's not public. We'd like functionalities in PyTorch's OSS code be accessible for all PyTorch users.
> 
> There are three solutions: (1) Advocate and upstream the API to NCCL; (2) Establish an abstract interface between PyTorch and a "plugin". In the plugin, you can make this non-public API call; (3) "Unpack" the functionality of the API here, for example, in this case: `hash = std::hash(ncclComm_.comm)`

I agree that either (1) or (2) are good in general but it will take time. We can take them asynchronously.  To unblock, would it be acceptable if we guard the use of ncclCommGetUniqueHash by using the above NCCL_COMM_GET_UNIQUE_HASH macro, and make it no-op when the macro is not defined? 
E.g.
```
uint64_t getCommHash() {
#ifdef NCCL_COMM_GET_UNIQUE_HASH
   ncclCommGetUniqueHash(comm, &commHash_);
#endif
}

Reply:
Hi @minsii thanks for the clarification.
> commHash which is a global unique id used by all ranks in a given NCCL communicator

I think `ncclUniqueId` serves this purpose. Can you get a has from it?
Per nccl.h, its structure is:
```
typedef struct { char internal[NCCL_UNIQUE_ID_BYTES]; } ncclUniqueId
```
so you can get a hash from `ncclUniqueId` or `ncclUniqueId.internal`.

Reply:
> Hi @minsii thanks for the clarification.
> 
> > commHash which is a global unique id used by all ranks in a given NCCL communicator
> 
> I think `ncclUniqueId` serves this purpose. Can you get a has from it? Per nccl.h, its structure is:
> 
> ```
> typedef struct { char internal[NCCL_UNIQUE_ID_BYTES]; } ncclUniqueId
> ```
> 
> so you can get a hash from `ncclUniqueId` or `ncclUniqueId.internal`.



> Hi @minsii thanks for the clarification.
> 
> > commHash which is a global unique id used by all ranks in a given NCCL communicator
> 
> I think `ncclUniqueId` serves this purpose. Can you get a has from it? Per nccl.h, its structure is:
> 
> ```
> typedef struct { char internal[NCCL_UNIQUE_ID_BYTES]; } ncclUniqueId
> ```
> 
> so you can get a hash from `ncclUniqueId` or `ncclUniqueId.internal`.

It would require us to expose the "hash" function inside NCCL to get the same hashed value at NCCL PG level v.s. NCCL internal. The goal for us to expose commHash is to have same id in NCCL PG and NCCL, which would help us quickly find corresponding log/trace cross stack.

Besides, uniqueId will not be different in the commSplit mode between parent comm v.s. child comm. `commHash` today also doesn't fully guarantee the uniqueness in such use case, but we can fix it in NCCL-EXP to make it unique. In contrast, it would be harder to support the commSplit case if we change to expose the hash function. Hope the design choice makes sense to you.

Reply:
I understand your use case better now.

I wonder if we can use the value in `ncclComm_t` to identify whether a PG's comm is the same as the one being operated in NCCL. And then use `ncclUniqueId` to identify whether a comm is the same one across ranks?
(The two together should solve the global + local identification need).

My very first comment holds whether there is #ifdef or not.
The #ifdef makes sure it builds, but does not achieve the concept of General Availability.

From code maintenance point of view, if different organizations have their own "NCCL" API definitions and all upstream to pytorch, then our code management would become intractable. This practice would be more acceptable under forks of pytorch, such as "mepa/pytorch", "geegle/pytorch", "macrosoft/pytorch" (all names made up), but less under "pytorch/pytorch."

Reply:
For this specific case: 

If a quick solution is needed, we can do (3):
- copy NCCL's internal hashing function here;
- write a comment that says "this is how NCCL hashes its communicator IDs in logs; to ease debugging, we prefer the same hash, thus copying NCCL's hash function here."

For a long-term solution in case this functionality is surely needed, I recommend contacting NCCL to create an access API.